New term entry on the built in function for Numpy named 'svd' (#4547)
Codecademy/docs: content/numpy/concepts/built-in-functions/terms/svd/svd.md

---
Title: '.svd()'

> Description: 'The `.svd()` built-in function performs the Singular Value Decomposition (SVD) on a matrix, breaking it down into singular vectors and values.'

Reviewer: ```suggestion
Description: 'Performs the Singular Value Decomposition (SVD) on a matrix, breaking it down into singular vectors and values.'
```

Subjects:
  - 'Computer Science'
  - 'Data Science'
  - 'Machine Learning'
Tags:
  - 'Linear Algebra'
  - 'Machine Learning'
  - 'NumPy'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
  - 'paths/data-science'
  - 'paths/data-science-foundaions'
---

The **`.svd()`** function is a mathematical technique that decomposes a matrix into three simpler matrices: it factorizes the matrix `a` into two unitary matrices `U` and `Vh`, along with a 1-D array `s` of singular values (real and non-negative). This decomposition satisfies the equation `a = U @ S @ Vh`, where `S` is a suitably shaped matrix of zeros with `s` as its main diagonal.

> ## Syntax

Reviewer: ```suggestion

## Syntax
```

```pseudo
numpy.linalg.svd(a, full_matrices=True, compute_uv=True, hermitian=False)
```

> `.svd()` provides the following arguments:

Reviewer: ```suggestion

`.svd()` provides the following arguments:
```

- `hermitian`: This parameter indicates whether the input matrix `a` is Hermitian, meaning it is equal to its conjugate transpose. When `hermitian` is set to `True`, the function assumes that a is Hermitian and uses a more efficient algorithm tailored for such matrices. If set to `False`, it uses a general algorithm. The default value is `False`.

## Example
The following example demonstrates various scenarios of Singular Value Decomposition (SVD) and uses `.svd()` using `NumPy`:

```py
# The following example demonstrates a scenario of Singular Value Decomposition (SVD) for a 2D matrix using NumPy:
import numpy as np

# Create a 2D array
A = np.random.randn(5, 3)  # Creating a random 2D matrix

print("Original 2D matrix:") 
print(A) 

# Compute the factor by Singular Value Decomposition
U, S, V = np.linalg.svd(A)  

# Print the result 
print("\nFactor of the given array by Singular Value Decomposition:") 
print("\nU=", U, "\n\ns=", S, "\n\nV=", V)

> ```

Reviewer: ```suggestion
```

## Output

```shell
Original 2D matrix:
[[ 0.24802525  0.08222177 -1.08276458]
 [ 1.69668935  0.4094856  -0.23407198]
 [ 0.40216763  2.30735008 -0.33489652]
 [-1.00632881  0.43372609 -1.97297003]
 [ 1.44445438  2.76595846  0.08637226]]

Factor of the given array by Singular Value Decomposition:

U= [[-0.06608409 -0.31521528  0.48785517 -0.65606804 -0.47733037]
 [-0.28695698  0.23340955  0.79622903  0.47670088  0.04417332]
 [-0.56145359 -0.20509892 -0.32859417  0.36975665 -0.63088049]
 [-0.01794366 -0.88995149  0.09929954  0.22345786  0.3845394 ]
 [-0.77313419  0.10990942 -0.10090663 -0.39456001  0.47362822]] 

s= [4.01680091 2.5291219  1.49878361] 

V= [[-0.45503019 -0.88743457  0.07353512]
 [ 0.50994021 -0.19198956  0.83851117]
 [ 0.73000582 -0.41904641 -0.53989962]]
```
```

Author: Dear @mamtawardhani , Thanks for your feedback. I put the output in a shell block, like the one has been done for the [numpy.sum()](https://github.com/Codecademy/docs/blob/main/content/numpy/concepts/built-in-functions/terms/sum/sum.md). Additionally, I've implemented the other suggested improvements.

import numpy as np

# Example 1: Full SVD, 2D Case
A = np.random.randn(5, 3)
U, S, Vt = np.linalg.svd(A, full_matrices=True)
U.shape, S.shape, Vt.shape
((5, 5), (3,), (3, 3))
np.allclose(A, np.dot(U[:, :3] * S, Vt))
True
Sigma = np.zeros((5, 3))
Sigma[:3, :3] = np.diag(S)
np.allclose(A, np.dot(U, np.dot(Sigma, Vt)))
True

# Example 2: Reduced SVD, 2D Case
B = np.random.randn(4, 6)
U, S, Vt = np.linalg.svd(B, full_matrices=False)
U.shape, S.shape, Vt.shape
((4, 4), (4,), (4, 6))
np.allclose(B, np.dot(U * S, Vt))
True
Sigma = np.diag(S)
np.allclose(B, np.dot(U, np.dot(Sigma, Vt)))
True

# Example 3: Full SVD, 3D Case
C = np.random.randn(3, 4, 5)
U, S, Vt = np.linalg.svd(C, full_matrices=True)
U.shape, S.shape, Vt.shape
((3, 3, 3), (3,), (3, 4, 5))
np.allclose(C, np.matmul(U[..., :3] * S[..., None, :], Vt))
True
np.allclose(C, np.matmul(U, np.matmul(np.diag(S), Vt)))
True

# Example 4: Reduced SVD, 3D Case
D = np.random.randn(2, 2, 3)
U, S, Vt = np.linalg.svd(D, full_matrices=False)
U.shape, S.shape, Vt.shape
((2, 2, 2), (2,), (2, 2, 3))
np.allclose(D, np.matmul(U * S[..., None, :], Vt))
True
np.allclose(D, np.matmul(U, np.matmul(np.diag(S), Vt)))

> True

Reviewer: Hey @doroodianMostafa, based on the entry you made, the `.svd()` computes the singular value decomposition and produces the `U', `Vh`, and `s` matrices, the example should be something similar to that. Please refer to the following links for reference:

https://www.geeksforgeeks.org/compute-the-factor-of-a-given-array-by-singular-value-decomposition-using-numpy/

https://python.quantecon.org/svd_intro.html

Please modify the example so that it fits your description.
Thanks!

Author: Hi @mamtawardhani,

Thanks for the feedback! I took a look at the links and tried to create a similar, but simpler example. To keep things clear, I skipped the 3D matrix example and left the other input parameters at their defaults.

---
Title: '.svd()'
Description: 'Singular Value Decomposition'
Subjects:
  - 'Computer Science'
  - 'Data Science'
  - 'Machine Learning'
Tags:
  - 'Linear Algebra'
  - 'Machine Learning'
  - 'NumPy'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
  - 'paths/data-science'
  - 'paths/data-science-foundaions'
---

The **`.svd()`** is a mathematical technique that decomposes a matrix into three simpler matrices.Factorizes the matrix `a` into two unitary matrices `U` and `Vh`, 
along with a 1-D array `s` of singular values (real and non-negative), such that `a` equals `U @ S @ Vh`, where `S` is a suitably shaped matrix of zeros with `s` as 
its main diagonal.

> 

Reviewer: Please remove the extra one line spacing here.

## Syntax

```pseudo
numpy.linalg.svd(a, full_matrices=True, compute_uv=True, hermitian=False)
```

> `.svd()` provides the following arguments:

Reviewer: This line can be removed.

> - `a`: This is the input matrix to be decomposed with a.ndim >= 2. It is the matrix you want to perform Singular Value Decomposition on.
> - `full_matrices`:This parameter determines whether to compute the full-sized or reduced-sized matrices U and VT. If full_matrices is set to True, the function 
> computes the full-sized matrices. If set to False, it computes only the essential parts of U and VT. By default, it's set to True.
> - `compute_uv`: This parameter specifies whether to compute the left-singular vectors (U) and right-singular vectors (transpose of V) in addition to the singular 
> values. If compute_uv is set to True, the function computes U and VT. If set to False, it only computes the singular values. By default, it's set to True.
> - `hermitian':This parameter indicates whether the input matrix a is Hermitian (equal to its conjugate transpose). If hermitian is set to True, the function assumes 
> that a is Hermitian and uses a more efficient algorithm tailored for Hermitian matrices. If set to False, it uses a general algorithm. By default, it's set to 
> False.

Reviewer: ```suggestion
- `a`: This parameter represents the input matrix to be decomposed, where `a.ndim>=2`. It is the matrix on which Singular Value Decomposition will be performed.
- `full_matrices`: This parameter determines whether the function computes full-sized or reduced-sized matrices `U` and `Vh`. If `full_matrices` is set to `True`, the function computes the full-sized matrices. If set to `False`, it computes only the essential parts of `U` and `Vh`. The default value is `True`.
- `compute_uv`: This parameter specifies whether the function computes the left-singular vectors (`U`) and right-singular vectors (transpose of `V`) in addition to the singular values. When `compute_uv` is set to `True`, the function computes `U` and `Vh`. If set to `False`, it only computes the singular values. The default value is `True`.
- `hermitian': This parameter indicates whether the input matrix `a` is Hermitian, meaning it is equal to its conjugate transpose. When `hermitian` is set to `True`, the function assumes that a is Hermitian and uses a more efficient algorithm tailored for such matrices. If set to `False`, it uses a general algorithm. The default value is `False`.

```

b = np.random.randn(2, 7, 8, 3) + 1j*np.random.randn(2, 7, 8, 3)

## Reconstruction based on full SVD, 2D case:

U, S, Vh = np.linalg.svd(a, full_matrices=True)
U.shape, S.shape, Vh.shape
((9, 9), (6,), (6, 6))
np.allclose(a, np.dot(U[:, :6] * S, Vh))
True
smat = np.zeros((9, 6), dtype=complex)
smat[:6, :6] = np.diag(S)
np.allclose(a, np.dot(U, np.dot(smat, Vh)))
True

## Reconstruction based on reduced SVD, 2D case:

U, S, Vh = np.linalg.svd(a, full_matrices=False)
U.shape, S.shape, Vh.shape
((9, 6), (6,), (6, 6))
np.allclose(a, np.dot(U * S, Vh))
True
smat = np.diag(S)
np.allclose(a, np.dot(U, np.dot(smat, Vh)))
True

## Reconstruction based on full SVD, 4D case:

U, S, Vh = np.linalg.svd(b, full_matrices=True)
U.shape, S.shape, Vh.shape
((2, 7, 8, 8), (2, 7, 3), (2, 7, 3, 3))
np.allclose(b, np.matmul(U[..., :3] * S[..., None, :], Vh))
True
np.allclose(b, np.matmul(U[..., :3], S[..., None] * Vh))
True

## Reconstruction based on reduced SVD, 4D case:

U, S, Vh = np.linalg.svd(b, full_matrices=False)
U.shape, S.shape, Vh.shape
((2, 7, 8, 3), (2, 7, 3), (2, 7, 3, 3))
np.allclose(b, np.matmul(U * S[..., None, :], Vh))
True
np.allclose(b, np.matmul(U, S[..., None] * Vh))

> True

Reviewer: Please modify or replace this example with a new one, this is plagiarized.
Plagiarized from: https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html

Author: Thanks for your comment. To clarify, I specifically mentioned the source of the examples in the commit comment to avoid plagiarism. However, I wasn't aware that it could still be considered plagiarism even when the resource is provided. Thanks for correcting me!
